### PR TITLE
JP-3598: The ResampleSpecStep and Extract1DStep within MasterBackgroundMosStep are inaccessible

### PR DESCRIPTION
**Pull request by penaguerrero**
Resolves [JP-3598](https://jira.stsci.edu/browse/JP-3598)
and 
Resolves [JP-3748](https://jira.stsci.edu/browse/JP-3748)

Closes #

This PR adds resample_spec and extract_1d as steps of the MOS background pipeline so that their respective step attributes can be modified there as well.

## Tasks
- [x] **request a review from someone specific**, to avoid making the maintainers review every PR
- [x] add a build milestone, i.e. `Build 11.3` (use the [latest build](https://github.com/spacetelescope/jwst/milestones) if not sure)
- [x] Does this PR change user-facing code / API? (if not, label with `no-changelog-entry-needed`)
  - [x] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see below for change types) 
  - [x] update or add relevant tests
  - [x] update relevant docstrings and / or `docs/` page
  - [x] [start a regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) and include a link to the running job ([click here for instructions](https://github.com/spacetelescope/RegressionTests/blob/main/docs/running_regression_tests.md))
    - [x] Do truth files need to be updated ("okified")?
      - [ ] **after the reviewer has approved these changes**, run `okify_regtests` to update the truth files
- [ ] if a JIRA ticket exists, [make sure it is resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

<details><summary>news fragment change types...</summary>

- ``changes/<PR#>.general.rst``: infrastructure or miscellaneous change
- ``changes/<PR#>.docs.rst``
- ``changes/<PR#>.stpipe.rst``
- ``changes/<PR#>.datamodels.rst``
- ``changes/<PR#>.scripts.rst``
- ``changes/<PR#>.fits_generator.rst``
- ``changes/<PR#>.set_telescope_pointing.rst``
- ``changes/<PR#>.pipeline.rst``

## stage 1
- ``changes/<PR#>.group_scale.rst``
- ``changes/<PR#>.dq_init.rst``
- ``changes/<PR#>.emicorr.rst``
- ``changes/<PR#>.saturation.rst``
- ``changes/<PR#>.ipc.rst``
- ``changes/<PR#>.firstframe.rst``
- ``changes/<PR#>.lastframe.rst``
- ``changes/<PR#>.reset.rst``
- ``changes/<PR#>.superbias.rst``
- ``changes/<PR#>.refpix.rst``
- ``changes/<PR#>.linearity.rst``
- ``changes/<PR#>.rscd.rst``
- ``changes/<PR#>.persistence.rst``
- ``changes/<PR#>.dark_current.rst``
- ``changes/<PR#>.charge_migration.rst``
- ``changes/<PR#>.jump.rst``
- ``changes/<PR#>.clean_flicker_noise.rst``
- ``changes/<PR#>.ramp_fitting.rst``
- ``changes/<PR#>.gain_scale.rst``

## stage 2
- ``changes/<PR#>.assign_wcs.rst``
- ``changes/<PR#>.badpix_selfcal.rst``
- ``changes/<PR#>.msaflagopen.rst``
- ``changes/<PR#>.nsclean.rst``
- ``changes/<PR#>.imprint.rst``
- ``changes/<PR#>.background.rst``
- ``changes/<PR#>.extract_2d.rst``
- ``changes/<PR#>.master_background.rst``
- ``changes/<PR#>.wavecorr.rst``
- ``changes/<PR#>.srctype.rst``
- ``changes/<PR#>.straylight.rst``
- ``changes/<PR#>.wfss_contam.rst``
- ``changes/<PR#>.flatfield.rst``
- ``changes/<PR#>.fringe.rst``
- ``changes/<PR#>.pathloss.rst``
- ``changes/<PR#>.barshadow.rst``
- ``changes/<PR#>.photom.rst``
- ``changes/<PR#>.pixel_replace.rst``
- ``changes/<PR#>.resample_spec.rst``
- ``changes/<PR#>.residual_fringe.rst``
- ``changes/<PR#>.cube_build.rst``
- ``changes/<PR#>.extract_1d.rst``
- ``changes/<PR#>.resample.rst``

## stage 3
- ``changes/<PR#>.assign_mtwcs.rst``
- ``changes/<PR#>.mrs_imatch.rst``
- ``changes/<PR#>.tweakreg.rst``
- ``changes/<PR#>.skymatch.rst``
- ``changes/<PR#>.exp_to_source.rst``
- ``changes/<PR#>.outlier_detection.rst``
- ``changes/<PR#>.tso_photometry.rst``
- ``changes/<PR#>.stack_refs.rst``
- ``changes/<PR#>.align_refs.rst``
- ``changes/<PR#>.klip.rst``
- ``changes/<PR#>.spectral_leak.rst``
- ``changes/<PR#>.source_catalog.rst``
- ``changes/<PR#>.combine_1d.rst``
- ``changes/<PR#>.ami.rst``

## other
- ``changes/<PR#>.wfs_combine.rst``
- ``changes/<PR#>.white_light.rst``
- ``changes/<PR#>.cube_skymatch.rst``
- ``changes/<PR#>.engdb_tools.rst``
- ``changes/<PR#>.guider_cds.rst``
</details>

### CODE REVIEW

**Comment by penaguerrero** [comment]
Reg tests running: https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1766/

**Review by melanieclarke — COMMENTED** [review]
Code changes look good to me.  I'm attempting to test now with some real data.

Can you please update the documentation for master background with MOS to add in the pixel_replace step?  Also, it looks like there is not yet any mention of how to set parameters in the substeps for the master_background_mos pipeline -- can we add a sentence or two about that to the description?

**Review by melanieclarke — CHANGES_REQUESTED** [review]
On closer look, and testing on real data, I think this needs more work.  Setting the other steps up as part of the step_defs allows it to take input parameters for the substeps, but then they are not actually used, since the steps are created and called in `create_background_from_multislit` instead of in the `process function`, like the other steps are.

I'm testing by just running spec2 from the command line on a MOS rate file with background slits, setting `--steps.master_background_mos.steps.pixel_replace.skip=False`.  The top-level parameters report that skip is set to False, but when the step is created later, the default parameter `skip=True` is used.

**Comment by penaguerrero** [comment]
Ok, step parameters are being updated now inside of create_background_from_multislit

**Comment by melanieclarke** [comment]
> Ok, step parameters are being updated now inside of create_background_from_multislit

That looks like it will almost work for passing the parameters from the top level pipeline to the point where the steps are called, but you're using the `call` class function to run the step, which creates a new instance of the step instead of running with current parameters.  This version is now making the step 3 times: once when it's established as part of the pipeline, once when it's directly instantiated to update the parameters, and once when `call` is called.

I think we can make the step just once, and avoid having to pass parameters around, if we move the step processing out of the `create_background_from_multislit` and call it directly in the `process` function for `MasterBackgroundMosStep`.  It needs to be called the same way the `barshadow` or `photom` steps are called.

**Comment by penaguerrero** [comment]
@melanieclarke this is the other alternative I had considered. I think it is cleaner.

**Comment by melanieclarke** [comment]
> @melanieclarke this is the other alternative I had considered. I think it is cleaner.

Thank you!  I'll take a look.

**Comment by melanieclarke** [comment]
>  this is the other alternative I had considered. I think it is cleaner.

Thanks for the updates - I agree, this looks better now!  

Two more suggestions:
1. I think we should combine the `_set_steps_params` with the already existing `set_pars_from_parent` function.  They are doing almost the same thing.
2. We might as well include combine_1d as part of the pipeline too, instead of handling it separately.  The user might want control over those parameters too.

**Comment by penaguerrero** [comment]
> 1. I think we should combine the `_set_steps_params` with the already existing `set_pars_from_parent` function.  They are doing almost the same thing.
I think it would be better to leave them in different functions since they do slightly different things, one sets the parameters the same as the parent step, whereas the other just uses the defaults regardless of the parent params

> 2. We might as well include combine_1d as part of the pipeline too, instead of handling it separately.  The user might want control over those parameters too.
I agree, I included it as part of this PR.

**Comment by melanieclarke** [comment]
> I think it would be better to leave them in different functions since they do slightly different things, one sets the parameters the same as the parent step, whereas the other just uses the defaults regardless of the parent params

Is there a reason not to use the parent params, the same way the other steps do?  It seems like that would still be the desired starting point for the additional steps.

**Review by melanieclarke — COMMENTED** [review]
I think this is close now - a couple small things left to tidy up, and then we should run regression tests.

@hayescr - these changes will conflict with yours in #8932.  Should we try to get this one in first and deconflict in your PR or the other way around?

File: changes/8847.refpix.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Include resample, pixel_replace, extract_1d, and combine_1d into MOS background pipeline
```
Comment (melanieclarke) [inline changes/8847.refpix.rst:1]: This change note is in the wrong place - it should be written to changes/8847.master_background.rst, not refpix

**Review by melanieclarke — COMMENTED** [review]
One more thing - this needs documentation updates, in docs/jwst/master_background/description.rst, under the NIRSpec MOS Mode heading.

**Comment by hayescr** [comment]
@melanieclarke - either way is fine with me.

I'll take a look at this PR to see if I have any feedback, but one thing I did notice was the switch to call the `combine_1d` step directly, and I want to get your all's thoughts on this.  The only parameter available for this step currently is the `exptime_key`, and all of the options should perform identically for the MOS background spectra (because all of the spectra have the same integration time and exposure time so the weighting will be the same for all spectra, similar to the unit weights option), so it's not really necessary to provide access to this option.

https://github.com/spacetelescope/jwst/pull/8932 will add another parameter for `combine_1d.combine_1d_spectra` that we will want to have different defaults when called by `master_background_mos `vs. the `combine_1d` step.  Doing so is straightforward if both steps call `combine_1d_spectra` with their own default settings, but it might be less straightforward if calling the `combine_1d` step in `master_background_mos`.  So I was wondering what you all thought of keeping the combine_1d_spectra function call instead of calling the step?

**Comment by melanieclarke** [comment]
> So I was wondering what you all thought of keeping the combine_1d_spectra function call instead of calling the step?

This was my suggestion/fault - I figured it would be more forward looking if the combine_1d call was part of the pipeline too. That way if we add more parameters to it in the future, the user can directly control them as needed.

For the case where you want master_background_mos to have a different default for specific combine_1d parameters, we could pass them directly from the top-level parameters to the step before calling it, e.g.
```
self.combine_1d.sigma_clip = self.sigma_clip
master_background = self.combine_1d.run(bkg_model)
```
I can see that potentially being confusing, though, if the user sets sigma_clip directly in the substep, and then it's overridden by the top-level parameters.  

Since combine1_d is currently a simple step, and there is already the self-contained `combine_1d_spectra` function, it would be fine by me to leave it as is.

**Comment by penaguerrero** [comment]
@melanieclarke and @hayescr I merged the changes of https://github.com/spacetelescope/jwst/pull/8932. Can you please take a look at the code?

File: jwst/master_background/master_background_mos_step.py [hunk 2]
```diff
@@ -62,6 +66,10 @@ class MasterBackgroundMosStep(Pipeline):
         'pathloss': pathloss_step.PathLossStep,
         'barshadow': barshadow_step.BarShadowStep,
         'photom': photom_step.PhotomStep,
+        'pixel_replace': pixel_replace_step.PixelReplaceStep,
+        'resample_spec': resample_spec_step.ResampleSpecStep,
+        'extract_1d': extract_1d_step.Extract1dStep,
+        'combine_1d': combine_1d_step.Combine1dStep
```
Comment (melanieclarke) [inline jwst/master_background/master_background_mos_step.py:72]: This should no longer be included here, since we'll call the function directly

**Review by hayescr — COMMENTED** [review]
I don't think I have any other comments on top of what Melanie mentioned and the changes work for me.  In terms of the tests, just in case it helps streamline adjusting the tests, I had set up a few of them to work around applying/removing some of the NIRSpec correction steps.  I just did this to avoid the effects of the pixel-to-pixel differences in the flat-fielding, which can make the values of the background and background correction difficult to predict ahead of time.

**Comment by penaguerrero** [comment]
Regtest: https://github.com/spacetelescope/RegressionTests/actions/runs/12147624590

**Review by melanieclarke — APPROVED** [review]
All comments addressed and regtest results look good - the only change is that the pixel replace keyword is added in the background spectra (and set to SKIPPED, as expected).

**Pull request merged**